spanner/batchwriter.go

```diff
@@ -0,0 +1,320 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package spanner provides high-level abstractions for working with
+// Cloud Spanner that are not available from the core Cloud Spanner
+// libraries.
+package spanner
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	sp "cloud.google.com/go/spanner"
+)
+
+// Parameters used to control building batches to write to Spanner.
+// Batches are built by adding rows until we hit one of the thresholds.
+// Bigger batches are usually more efficient, but we need to be careful
+// not to exceed Spanner's limits. Also, sending huge RPCs is potentially
+// unreliable.
+const (
+	kb             = 1 << 10
+	mb             = 1 << 20
+	countThreshold = 10 * kb // Spanner per-operation limit is 20K.
+	byteThreshold  = 20 * mb // Spanner per-operation limit is 100MB.
+)
+
+// BatchWriter accumulates Spanner mutations (via AddRow) and assembles
+// them into batches that it asynchronously writes to Spanner. If Spanner
+// returns an error for a batch, BatchWriter splits the batch into
+// smaller chunks to retry, as it attempts to isolate which row(s) in a
+// batch is bad. BatchWriter respects Spanner's limits on byte size and
+// mutation count and has configurable limits on the number of
+// in-progress writes, amount of data buffered and retry behavior.
+// BatchWriter is not threadsafe: only one call to AddRow or Flush should
```

Comment:
It would be a moderate amount of work. All access to rows, rBytes, rCount would need to be either protected by locks or use atomics. Somewhat more code and complexity. Also, the current locking is not on the critical path (its only for handling errors). But to make the library thread-safe, we'd need to protect rows, which is in the critical path. It might make sense to use go channels instead i.e. there's probably some thinking to do to make this change.

Slight preference for not doing it now since it isn't needed for harbourbridge. I think the trigger to do it would be if there is interest in using this code elsewhere.